src/TheNextWeek/planar.h

```diff
@@ -0,0 +1,158 @@
+#ifndef PLANAR_H
+#define PLANAR_H
+//==============================================================================================
+// To the extent possible under law, the author(s) have dedicated all copyright and related and
+// neighboring rights to this software to the public domain worldwide. This software is
+// distributed without any warranty.
+//
+// You should have received a copy (see file COPYING.txt) of the CC0 Public Domain Dedication
+// along with this software. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//==============================================================================================
+
+#include "rtweekend.h"
+
+#include "hittable.h"
+
+#include <cmath>
+
+//// This class is somewhat complex, but that's because it serves as a base class for very
+//// simple derived classes, where you really only need to worry about construction, bounds,
+//// and UV hit determination.
+
+class planar : public hittable {
+  public:
+    planar(const point3& o, const vec3& side_A, const vec3& side_B, shared_ptr<material> m)
+      : origin(o), a(side_A), b(side_B), mat(m)
+    {
+        normal = unit_vector(cross(a, b));
+        D = -dot(normal, origin);
+        a /= dot(a,a);
+        b /= dot(b,b);
+
+/////// By default, bounded to the world coordinates of UV<0,0> to UV<1,1>.
+
+        interval u_bounds, v_bounds;
+        get_uv_bounds(u_bounds, v_bounds);
+        bbox = aabb(
+            origin + (u_bounds.min * side_A) + (v_bounds.min * side_B),
+            origin + (u_bounds.max * side_A) + (v_bounds.max * side_B)
+        );
+    }
+
+    virtual void get_uv_bounds(interval& u_bounds, interval& v_bounds) {
+        u_bounds = interval::unit;
+        v_bounds = interval::unit;
+    }
+
+    virtual bool hit_uv(double u, double v) const = 0;
+
+/// The distinction of this method is that it calls `hit_uv()` to filter on UV coordinates.
+
+    bool hit(const ray& r, interval ray_t, hit_record& rec) const override {
+        auto denom = dot(normal, r.direction());
+        if (fabs(denom) < 1e-8) return false;
+
+        auto t = (-D - dot(normal, r.origin())) / denom;
+        if (!ray_t.contains(t))
+            return false;
+
+        auto intersection = r.at(t);
+
+        vec3 planar_hitpt_vector = intersection - origin;
+
+        auto u = dot(planar_hitpt_vector, a);
+        auto v = dot(planar_hitpt_vector, b);
+        if (!hit_uv(u,v))
+            return false;
+
+        rec.t = t;
+        rec.p = intersection;
+        rec.u = u;
+        rec.v = v;
+        rec.mat = mat;
+        rec.set_face_normal(r, normal);
+
+        return true;
+    }
+
+    aabb bounding_box() const { return bbox; }
+
+  protected:
+    point3 origin;
```

Comment:
my brain defaulted to the origin as `(0,0,0)` which is not the case in general. Made parsing the code more difficult.
Consider renaming to something else.

Center? Modulus? Nexus? Terminus? I don't really have anything I particularly like

Reply:
Well, it _is_ `(0,0)`. Certainly warrants a comment though. The basic idea is that we're defining a new 2D coordinate system (or frame). Given that, "origin" is the most accurate term, but I can improve the naming. Something like "plane_origin" or "origin_2D".

Reply:
While I agree with you, if we're defining a new plane (which I agree with, but since there isn't associated text, that's not where my brain went with), then `vec3 a`, and `vec3 b` should actually probably be `vec3 x` and `vec3 y/z`.
But that would be maximally confusing.

maybe `point3 plane_origin`, `vec3 plane_x`, `vec3 plane_y`.

That depends if we want the normal of the plane to `y` or `z`. Which, I could genuinely be argued into either.